src/ekka_rlog_replica.erl

```diff
@@ -0,0 +1,311 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2021 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+%% @doc This module implements a gen_statem which collects rlogs from
+%% a remote core node.
+-module(ekka_rlog_replica).
+
+%% API:
+-export([start_link/1, push_batch/3]).
+
+%% gen_statem callbacks:
+-export([init/1, terminate/3, code_change/4, callback_mode/0, handle_event/4]).
+
+-include_lib("snabbkaffe/include/snabbkaffe.hrl").
+
+%%================================================================================
+%% Type declarations
+%%================================================================================
+
+%% States:
+-define(disconnected, disconnected).
+-define(bootstrap, bootstrap).
+-define(local_replay, local_replay).
+-define(normal, normal).
+
+-type state() :: ?bootstrap
+               | ?local_replay
+               | ?normal
+               | ?disconnected.
+
+-record(d,
+        { shard                        :: ekka_rlog:shard()
+        , remote_core_node = undefined :: node() | undefined
+        , agent                        :: pid() | undefined
+        , tmp_worker       = undefined :: pid() | undefined
+        , checkpoint       = undefined :: ekka_rlog_server:checkpoint()
+        , next_batch_seqno = 0         :: integer()
+        }).
+
+-type data() :: #d{}.
+
+-type fsm_result() :: gen_statem:event_handler_result(state()).
+
+%%================================================================================
+%% API funcions
+%%================================================================================
+
+%% This function is called by the remote core node.
+-spec push_batch(node(), ekka_rlog:shard(), ekka_rlog_lib:batch()) -> ok.
+push_batch(Node, Shard, Batch) ->
+    ekka_rlog_lib:rpc_call(Node, gen_statem, call, [Shard, {tlog_batch, Batch}, infinity]).
+
+start_link(Shard) ->
+    Config = #{}, % TODO
+    gen_statem:start_link(?MODULE, {Shard, Config}, []).
+
+%%================================================================================
+%% gen_statem callbacks
+%%================================================================================
+
+callback_mode() -> [handle_event_function, state_enter].
```

Comment:
could you elaborate why `handle_event_vunction` but not `state_functions`   ?

Reply:
I mean as code comment.

Reply:
Done, thanks for suggestion